src/clj/game/cards/ice.clj

```diff
@@ -889,6 +889,39 @@
    "Galahad"
    (grail-ice end-the-run)
 
+   "Gatekeeper"
+   (let [draw {:async true
+               :prompt "Draw how many cards?"
+               :choices {:number (req 3)
+                         :max (req 3)
+                         :default (req 1)}
+               :msg (msg "draw " target "cards")
+               :effect (effect (draw eid target nil))}
+         reveal-and-shuffle {:async true
+                             :prompt "Reveal and shuffle up to 3 agendas"
+                             :show-discard true
+                             :choices {:req #(and (= "Corp" (:side %))
+                                                  (or (= [:discard] (:zone %))
+                                                      (= [:hand] (:zone %)))
+                                                  (is-type? % "Agenda"))
+                                       :max (req 3)}
+                             :effect (req (doseq [c targets]
+                                            (move state :corp c :deck))
+                                          (shuffle! state :corp :deck)
+                                          (effect-completed state side eid))
+                             :cancel-effect (effect (shuffle! :deck)
+                                                    (effect-completed eid))
+                             :msg (msg "add "
+                                       (str (join ", " (map :title targets)))
+                                       " to R&D")}
+         draw-reveal-shuffle {:async true
+                              :label "Draw cards, reveal and shuffle agendas"
+                              :effect (req (wait-for (resolve-ability state side draw card nil)
+                                                     (continue-ability state side reveal-and-shuffle card nil)))}]
+    {:strength-bonus (req (if (= :this-turn (:rezzed card)) 6 0))
```

Comment:
Is this already part of the engine? Dayum.

Reply:
I know! I was also pleasantly surprised!